RFC: Remove `subpath` URLQueryParam from settings pages.

tomalec opened the pull request:
_(This is a follow-up from https://github.com/woocommerce/google-listings-and-ads/pull/1025#discussion_r717420234)_

### Changes proposed in this Pull Request:

Remove `subpath` URLQueryParam from settings pages.
Use WC navigation features instead.

This way we have 
- routing unified with the WC stack
- less noise in URLs & track events
- more intuitive URL & track event props, as the entire path is in a single param, not divided into a number of chunks
- could get closer to [simplify `.~/utils/url.js`](https://github.com/woocommerce/google-listings-and-ads/issues/1024), as we no longer need to construct a query object, but just use `path` param.
- remove the source of problems like the one discussed at https://github.com/woocommerce/google-listings-and-ads/pull/1025#discussion_r717420234, TL;DR: "what if `subpath` is empty?"


What's left/missing/to be done later:
- [ ] Update track event README
- [ ] Try to simplify `wc_admin_register_page` calls in `src/Menu/Settings.php`
- [ ] :scroll: extrapolate for all other pages using `subpath` or similar param.
- (edit) ~[ ] 📜(not sure if possible in WC-plaform) DRY/reduce duplication, of pages configs in `src/Menu/*.php`s, `js/src/index.js` and `js/src/utils/url.js`~ -> https://github.com/woocommerce/google-listings-and-ads/issues/1037


### Screenshots:

![path](https://user-images.githubusercontent.com/17435/135463973-1ab3084a-9067-4e4d-8514-682ec47cbafc.gif)

### Detailed test instructions:

1. Go to Settings page [`/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings`](https://gla1.test/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fsettings)
2. Navigate into phone & address pages - "Edit" buttons
3. Navigate away and around observing how URL changes, and what track events are fired.
(full page reloads instead of SPA-morph, is an unrelated issue https://github.com/woocommerce/google-listings-and-ads/issues/1027)


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Make all settings pages use only `path` for navigation.

-----

//cc @eason9487 @ecgan

tomalec committed:
Remove subpath URLQueryParam from settings pages.

Use WC navigation features instead.

ecgan commented:
@tomalec , the reason why we use `subpath` is to make the pages work with the new WC Navigation beta feature. Refer to the following links:

- https://github.com/woocommerce/google-listings-and-ads/pull/605#issuecomment-847288673
- https://github.com/woocommerce/google-listings-and-ads/issues/665
- https://github.com/woocommerce/google-listings-and-ads/pull/743 (This has video demo.)

You can enable the WC Navigation feature in WC Settings > Advanced > Features.

tomalec commented:
I managed to achieve that by using [`matchExpression`](https://woocommerce.github.io/code-reference/classes/Automattic-WooCommerce-Admin-Features-Navigation-Menu.html#method_add_plugin_item) property 
![image](https://user-images.githubusercontent.com/17435/136057389-a52cbe6c-2a3d-4fb0-b994-8742b4ea9823.png)

It was quite hidden in the docs, so I found it by debugging it live :man_facepalming: 
But I hope it's stable enough and desired usage.

tomalec committed:
Make settings subpages highlight Settings menu item,

in the new WC Navigation.

Addresses https://github.com/woocommerce/google-listings-and-ads/pull/1031#issuecomment-931503228.

tomalec commented:
Highlighting the subpages in legacy menu is covered by https://github.com/woocommerce/google-listings-and-ads/pull/1038

eason9487 commented on js/src/settings/index.js:
```diff
@@ -17,8 +17,15 @@ import EditStoreAddress from './edit-store-address';
+ * Settings page component.
+ *
+ * @param {Object} [props] React props
+ * @param {Object} [props.params = {}] `path` params extracted by React `<Router>`.
+ * @param {string} [props.params.subpath] Sub path of a nested settings page.
```
😎 👍

eason9487 commented on src/Menu/Settings.php:
```diff
@@ -29,9 +29,38 @@ function() {
 						'nav_args' => [
 							'order'  => 40,
 							'parent' => 'google-listings-and-ads-category',
+							// Highlight this menu item for other subpages.
+							'matchExpression' => '/google/settings(/[^/]+)?',
```
TIL: `matchExpression`. Great find! 🙌

eason9487 reviewed (approved):
Tested with enabled/disabled the Woo Navigation feature and works well. LGTM.

findingsimple commented:
Mental note - before we go merging will this impact any funnels we have set up or direct links we may have (in the codebase or docs)

tomalec commented:
> will this impact any funnels we have set up or direct links we may have (in the codebase or docs)

Yes, it will. All `path=/google/settings&subpath={subpath}` will become `path=/google/settings/{subpath}`.

So the `url`, `path` and `subpath` event props will change respectively.

That's why I still have on my todo list:
> Update track event README

tomalec commented:
@jconroy does the above make it a no-go? or maybe it's ok, but we should change everything at once, to update the docs and funnels only once, to minimize the inconvenience, for users of those.

eason9487 commented on src/API/Site/Controllers/Google/AccountController.php:
```diff
@@ -88,7 +88,7 @@ protected function get_connect_callback(): callable {
 			try {
 				$next = $request->get_param( 'next' );
-				$path = $next === 'setup-mc' ? '/google/setup-mc' : '/google/settings&subpath=/reconnect-accounts';
+				$path = $next === 'setup-mc' ? '/google/setup-mc' : '/google/settings/reconnect-accounts';
```
📝 
Will conflict with [this commit](https://github.com/woocommerce/google-listings-and-ads/commit/72d86706f1ffb6d6976508f56e3462dd150f7ad5#diff-699dc48ef5ba31467d95145db88b190b3466fc6e935d85117262bbc5cdd72c0bR33-R103) in PR #1069

eason9487 commented:
Thanks for proposing the RFC, @tomalec! After several years of iteration, the new WooCommerce Navigation has been deprecated (see #2406). The stale PRs related to it will likely require reassessment for appropriate solutions, so I am closing this PR.
